### ISSUE TITLE
[XLA] code generation for ARM NEON produce very slow code with tensorflow 1.9.0 and 1.10.0

### ISSUE THREAD

**Issue by guillaume-michel** [tensorflow/tensorflow#21761]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: NO
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 14.04
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**: Cortex A7
- **TensorFlow installed from (source or binary)**: from source
- **TensorFlow version (use command below)**: v1.9.0 and superior
- **Python version**: 3.4
- **Bazel version (if compiling from source)**: 0.15.2
- **GCC/Compiler version (if compiling from source)**: 4.9.4
- **CUDA/cuDNN version**: N/A
- **GPU model and memory**: N/A
- **Exact command to reproduce**:
tfcompile --graph=network.pb --config=network.config.pbtxt  --cpp_class=Network --out_header=network.h --out_function_object=network.o --xla_enable_fast_math=true --target_triple=armv7a-none-android --target_features=+neon --entry_point=run_network

arm-linux-androideabi-g++ -shared -o test.so network.o

(details of network.pb and network.config.pbtxt are not relevant, The only thing I can say is that it contains 2D Convolution and Dense layer)

### Describe the problem
I use XLA AOT (tfcompile) to compile network down to shared library with tfcompile and it was working great until v1.9.0.
Starting from v1.9.0, I see huge speed penalty (more than 15x slowdown) when executing the network on ARM v7 + NEON. After some investigation, **I think it is pretty clear that v1.9.0 and v1.10.0 do not use specialized neon runtime for matmul and conv** (even when +neon is specified in target_features) as shown in the following objdumps:

- ** With tensorflow v1.8.0 and previous versions
objdump -CT test.so               

test.so:     file format elf32-little

DYNAMIC SYMBOL TABLE:
00000000      DF *UND*	00000000  LIBC        __cxa_finalize
00000000      DF *UND*	00000000  LIBC        __cxa_atexit
00000000      D  *UND*	00000000              **__xla_cpu_runtime_EigenConvF32**
00000000      D  *UND*	00000000              **__xla_cpu_runtime_EigenMatMulF32**
000003c0 g    DF .text	0000abc0  Base        run_network
00672004 g    D  *ABS*	00000000  Base        __bss_start
00672004 g    D  *ABS*	00000000  Base        _end
00672004 g    D  *ABS*	00000000  Base        _edata

- ** With tensorflow v1.9.0 and superior:

objdump -CT test.so 

test.so:     file format elf32-little

DYNAMIC SYMBOL TABLE:
00000000      DF *UND*	00000000  LIBC        __cxa_finalize
00000000      DF *UND*	00000000  LIBC        __cxa_atexit
00000320 g    DF .text	0000cb40  Base        run_network
00674004 g    D  *ABS*	00000000  Base        __bss_start
00674004 g    D  *ABS*	00000000  Base        _end
00674004 g    D  *ABS*	00000000  Base        _edata

I tried to follow the code generation process but got lost in LLVM code generation.
Could you have a look at the problem, please? Any help will be helpful.

### Source code / logs
Running code just call run_network function in the shared library with the correct parameters.

**Comment by sanjoy** [comment]
Sorry for the regression.

Can you try changing `DotOpEmitter::EmitLlvmIrDotIfProfitable` to return `false` unconditionally and see if the regression disappears?  If yes the fix should be relatively simple.

**Comment by guillaume-michel** [comment]
@sanjoy Thanks for taking some time to investigate. I really appreciate.

I made the suggested change but XLA is still very slow (same as before).
Details:
I have the same objdump as before:
objdump -CT test.so

test.so: file format elf32-little

DYNAMIC SYMBOL TABLE:
00000000 DF UND	00000000 LIBC __cxa_finalize
00000000 DF UND	00000000 LIBC __cxa_atexit
00000320 g DF .text	0000cb40 Base run_network
00674004 g D ABS	00000000 Base __bss_start
00674004 g D ABS	00000000 Base _end
00674004 g D ABS	00000000 Base _edata

For reference I made a test that compare TF, XLA and TFLite on the same network and I have the following results on ARM with tensorflow 1.10:
- TF: 201 ms
- TFLite: 273 ms
- XLA: 4446 ms
Neural networks outputs for each inference engine are the same so correctness is not an issue and it indicates the same operations are performed for each tests.

I am happy to help if you have other tests I can perform.

**Comment by sanjoy** [comment]
Can you please attach `network.pb`  and `network.config.pbtxt` so that I can try reproducing this?

**Comment by guillaume-michel** [comment]
I am sorry but I can't provide that kind of information.
It is not restricted to this particular architecture, we observe the same regression for all our networks.

**Comment by sanjoy** [comment]
Can you extract out a small reproducer that you _can_ share?

**Comment by tensorflowbutler** [comment]
Nagging Assignees @sanjoy, @hawkinsp: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by mibrahimali** [comment]
I noticed the same issue , i`m trying to deploy Inception model on arm target , i managed to deployed for linux and Xeon processor but when i move to arm aarch64 it took almost 10X running time slower.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @sanjoy, @hawkinsp: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @sanjoy: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @sanjoy: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @sanjoy: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by sanjoy** [comment]
Hi,

I'm not actively working on this, but I'd be happy to review pull requests if someone (outside Google or within Google) wants to drive performance on aarch64.

**Comment by depthwise** [comment]
So as far as I can tell, in the ideal situation this is supposed to call into a platform specific runtime for the target platform to make use of highly optimized ops such as gemm, gemv, fft, etc. Where would it get that runtime when it's cross-compiling from Intel? Is there a high level doc I could read for how this is supposed to work?

**Comment by jvishnuvardhan** [comment]
I am closing this issue. Please open a new ticket if the issue persists with latest version of TFlow. Thanks!
